File: 1518/WaterBottles.ts
/*
https://leetcode.com/problems/water-bottles/description/?envType=daily-question&envId=2024-07-07
*/

function numWaterBottles(numBottles: number, numExchange: number): number {
    let maxNumberBottlesToDrink:number=0, currentExtraNumberBottlesToDrink:number=0, currentEmptiedBottles:number=0;
    maxNumberBottlesToDrink+=numBottles, currentEmptiedBottles+=numBottles;
    while(true){
        if(currentEmptiedBottles<numExchange) break;
        let recentEmptiedBottles:number=0;
        while(currentEmptiedBottles>=numExchange){
            currentExtraNumberBottlesToDrink++;
            currentEmptiedBottles-=numExchange;
        }
        recentEmptiedBottles+=currentEmptiedBottles+currentExtraNumberBottlesToDrink;
        maxNumberBottlesToDrink+=currentExtraNumberBottlesToDrink;
        currentExtraNumberBottlesToDrink=0;
        currentEmptiedBottles=recentEmptiedBottles;
    }
    return maxNumberBottlesToDrink;
};

File: 1/1000s/15s/1518/WaterBottles.cpp
/*
https://leetcode.com/problems/water-bottles/description/?envType=daily-question&envId=2024-07-07
*/

class Solution {
public:
    int numWaterBottles(int numBottles, int numExchange) {
        int res=0, nwres=0, curempty=0;
        res+=numBottles, curempty+=numBottles;
        while(true){
            if(curempty<numExchange) break;
            int nwempty=0;
            while(curempty>=numExchange){
                nwres++;
                curempty-=numExchange;
            }
            nwempty+=curempty;
            nwempty+=nwres;
            res+=nwres;
            nwres=0;
            curempty=nwempty;
        }
        return res;
    }
};